event.go

```diff
@@ -517,101 +711,239 @@ func checkID(id, kind string, sigil byte) (domain string, err error) {
+		eventJSON := e.eventJSON
+		var reference EventReference
+		reference, err = referenceOfEvent(eventJSON, e.roomVersion)
+		if err != nil {
+			return
+		}
+		eventID = reference.EventID
+	default:
+		err = errors.New("gomatrixserverlib: unknown room version")
+	}
+	return
+}
 
 // EventID returns the event ID of the event.
-func (e Event) EventID() string {
-	return e.fields.EventID
+func (e *Event) EventID() string {
+	switch fields := e.fields.(type) {
+	case eventFormatV1Fields:
+		return fields.EventID
+	case eventFormatV2Fields:
+		return fields.EventID
+	default:
+		panic(e.invalidFieldType())
+	}
 }
 
 // Sender returns the user ID of the sender of the event.
-func (e Event) Sender() string {
-	return e.fields.Sender
+func (e *Event) Sender() string {
+	switch fields := e.fields.(type) {
+	case eventFormatV1Fields:
+		return fields.Sender
+	case eventFormatV2Fields:
+		return fields.Sender
+	default:
+		panic(e.invalidFieldType())
+	}
 }
 
 // Type returns the type of the event.
-func (e Event) Type() string {
-	return e.fields.Type
+func (e *Event) Type() string {
+	switch fields := e.fields.(type) {
+	case eventFormatV1Fields:
+		return fields.Type
+	case eventFormatV2Fields:
+		return fields.Type
+	default:
+		panic(e.invalidFieldType())
+	}
 }
 
 // OriginServerTS returns the unix timestamp when this event was created on the origin server, with millisecond resolution.
-func (e Event) OriginServerTS() Timestamp {
-	return e.fields.OriginServerTS
+func (e *Event) OriginServerTS() Timestamp {
+	switch fields := e.fields.(type) {
+	case eventFormatV1Fields:
+		return fields.OriginServerTS
+	case eventFormatV2Fields:
+		return fields.OriginServerTS
+	default:
+		panic(e.invalidFieldType())
+	}
 }
 
 // Unsigned returns the object under the 'unsigned' key of the event.
-func (e Event) Unsigned() []byte {
-	return []byte(e.fields.Unsigned)
+func (e *Event) Unsigned() []byte {
+	switch fields := e.fields.(type) {
+	case eventFormatV1Fields:
+		return fields.Unsigned
+	case eventFormatV2Fields:
+		return fields.Unsigned
+	default:
+		panic(e.invalidFieldType())
+	}
 }
 
 // Content returns the content JSON of the event.
-func (e Event) Content() []byte {
-	return []byte(e.fields.Content)
+func (e *Event) Content() []byte {
+	switch fields := e.fields.(type) {
+	case eventFormatV1Fields:
+		return []byte(fields.Content)
+	case eventFormatV2Fields:
+		return []byte(fields.Content)
+	default:
+		panic(e.invalidFieldType())
+	}
 }
 
 // PrevEvents returns references to the direct ancestors of the event.
-func (e Event) PrevEvents() []EventReference {
-	return e.fields.PrevEvents
+func (e *Event) PrevEvents() []EventReference {
+	switch fields := e.fields.(type) {
+	case eventFormatV1Fields:
+		return fields.PrevEvents
+	case eventFormatV2Fields:
+		var result []EventReference
+		for _, id := range fields.PrevEvents {
+			result = append(result, EventReference{
+				EventID:     id,
+				EventSHA256: Base64String(id[1:]),
```

Comment:
Why `1:`?

Reply:
I'll comment this but mostly this is because the event ID in v3 onwards is actually the event hash. `1:` excludes the `$` sigil.